### PR TITLE
Element: Provide createHigherOrderComponent helper

### PR DESCRIPTION
**Pull request by aduth**
Context: https://github.com/WordPress/gutenberg/pull/5206#discussion_r170629217

This pull request seeks to experiment with an alternative to `@wordpress/element`'s `getWrapperDisplayName` function which is used to generate a display name for higher-order components. The use of this function is not always obvious, and can overcomplicate otherwise simple higher-order components by requiring the separate assignment of a class/function before then assigning `displayName`. The idea here is to eliminate consideration of the `displayName` with a generic helper for creating higher-order components, handling the assignment of the generated display name within the helper itself.

__Implementation notes:__

The code functionally works, though there are many other higher-order components we'd want to port if wanting to proceed, plus deprecating `getWrapperDisplayName`, or alternatively discouraging its use and leveraging it as the internal behavior of `createHigherOrderComponent`.

### CODE REVIEW

File: element/index.js
Comment (gziolo) [inline element/index.js]: We had `upperFirst` & `camelCase` on top of the modifier name to make sure it always looks like the name of the component. We may enforce it at the review time. Your call.

**Review by gziolo — APPROVED** [review]
This is what we exactly what we discussed some time ago. I like this new HOC helper. We should refactor also all the existing occurrences and deprecate `getWrapperDisplayName`. We should also adjust the existing test for `getWrapperDisplayName` to work with the HOC or provide something similar. I can help with that tomorrow.

**Comment by gziolo** [comment]
Yes, just wanted to say that I like how those HOCs evolve and make adding new functionalities easy to compose 💯

**Comment by gziolo** [comment]
@aduth - I added the following changes with d9bbcf9df0e295771fd2887916558af60517031c:
* Covered `createHigherOrderComponent` with unit tests.
* Added deprecation for `getWrapperDisplayName `.
* Refactored all occurences of `getWrapperDisplayName` with `createHigherOrderComponent`.

I tried to make this PR small so in some cases we might perform more refactorings, but I wanted to avoid too much noise to make the review process easier.

**Comment by gziolo** [comment]
Something to think a bit more. I noticed that in many cases we define the same name for `modifierName ` as the wrapping function. See:
```js
export const withAlign = createHigherOrderComponent( ( BlockListBlock ) => {
	return ( props ) => {
		...
		return <BlockListBlock { ...props } wrapperProps={ wrapperProps } />;
	};
}, 'withAlign' );
```

I'm not sure if it is a good idea, but we might want to make the `modifierName` optional. In such case, we could take it directly from the wrapper function's name. Something to think about, as a further improvement. It might be also confusing and error-prone.

File: element/index.js
Comment (gziolo) [inline element/index.js]: I replicated the old behavior `upperFirst( camelCase( modifierName ) )` - feel free to update if you think it's not necessary.

File: lib/client-assets.php
Comment (gziolo) [inline lib/client-assets.php]: I hope it's temporary until we remove deprecated function :)

File: components/higher-order/with-state/index.js
Comment (mcsf) [inline components/higher-order/with-state/index.js]: Missing `'withState'` as `modifierName`.

File: element/index.js
Comment (mcsf) [inline element/index.js]: Since we don't attempt to infer `modiferName`, should there be a warning if the argument is missing?

**Review by mcsf — APPROVED** [review]
Pending my observations, LGTM

File: components/higher-order/with-state/index.js
Comment (mcsf) [reply components/higher-order/with-state/index.js]: (amended & pushed)

Comment (gziolo) [reply components/higher-order/with-state/index.js]: Thanks 👍

**Comment by aduth** [comment]
>In such case, we could take it directly from the wrapper function's name. Something to think about, as a further improvement. It might be also confusing and error-prone.

Perhaps. I expect this wouldn't work so well for minified code (related: https://github.com/Automattic/wp-calypso/issues/18650 ). Whether we care about this is another question, as we're not really using the name in runtime environments except maybe React DevTools.

File: components/higher-order/with-state/index.js
Comment (aduth) [reply components/higher-order/with-state/index.js]: I don't see amend?

**Comment by gziolo** [comment]
You can instrument uglify to preserve function names, but still it might not always work. It seems to be okay as it is in that context.

File: components/higher-order/with-state/index.js
Comment (mcsf) [reply components/higher-order/with-state/index.js]: Oh. Maybe it got nuked in a subsequent rebase, @gziolo?

Comment (gziolo) [reply components/higher-order/with-state/index.js]: Yes, I did something wrong :(
Do you have this branch locally with your commit?

Comment (mcsf) [reply components/higher-order/with-state/index.js]: > Yes, I did something wrong

My fault for rebasing your commit. ;)

> Do you have this branch locally?

I do, the [rebase](https://gist.github.com/mcsf/cfb44ce3b5f3d9d39865c4d42a24184b) is `0ae3d2` and the [fix](https://gist.github.com/mcsf/0b28501c52fab9bb0b0a3f3c57a7ec4b) is `e3a846`.

Comment (gziolo) [reply components/higher-order/with-state/index.js]: Fixes with https://github.com/WordPress/gutenberg/commit/66b226898e3edc74a2a553c56648ff277002c0bd, thanks!

**Pull request merged**
